examples/advanced_operations/add_responsive_search_ad_full.py

```diff
@@ -0,0 +1,574 @@
+#!/usr/bin/env python
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This example shows how to create a complete Responsive Search ad.
+
+Includes creation of: budget, campaign, ad group, ad group ad,
+keywords, geo targeting, and image extensions.
+
+More details on Responsive Search ads can be found here:
+https://support.google.com/google-ads/answer/7684791
+"""
+
+import argparse
+import sys
+import uuid
+import requests
+
+from google.ads.googleads.client import GoogleAdsClient
+from google.ads.googleads.errors import GoogleAdsException
+
+
+# Keywords from user.
+_KEYWORD_TEXT_EXACT_1 = "example of exact match"
+_KEYWORD_TEXT_PHRASE_1 = "example of phrase match"
+_KEYWORD_TEXT_BROAD_1 = "example of broad match"
+
+# Geo targeting from user.
+_GEO_LOCATION_1 = "Buenos aires"
+_GEO_LOCATION_2 = "San Isidro"
+_GEO_LOCATION_3 = "Mar del Plata"
+
+# _LOCALE and _COUNTRY_CODE are used for geo targeting.
+# _LOCALE is using ISO 639-1 format. If an invalid _LOCALE is given,
+# 'es' is used by default.
+_LOCALE = "es"
+
+# A list of country codes can be referenced here:
+# https://developers.google.com/google-ads/api/reference/data/geotargets
+_COUNTRY_CODE = "AR"
+
+# To add image extensions, the account has to follow these requirements:
+# https://support.google.com/google-ads/answer/9566341
+# If the account does not meet the requirements, set below variable to False.
+_ADD_IMAGE_EXTENSION = True
+
+"""Support function"""
+
+
+def create_ad_text_asset(client, text, pinned_field=None):
+    """Create an AdTextAsset."""
+    ad_text_asset = client.get_type("AdTextAsset")
+    ad_text_asset.text = text
+    if pinned_field:
+        ad_text_asset.pinned_field = pinned_field
+    return ad_text_asset
+
+
+"""6 steps to create a Responsive Search Ad (RSA)."""
+
+
+def main(client, customer_id):
+    """
+  The main method that creates all necessary entities for the example.
+
+  Args:
+      client: an initialized GoogleAdsClient instance.
+      customer_id: a client customer ID.
+
+  Returns:
+      A responsive search ad with all settings required to run campaign.
+  """
+    # Step 1.1 -- Create budget.
+    # Create a budget, which can be shared by multiple campaigns.
+    campaign_budget = _create_campaign_budget(client, customer_id)
```

Comment:
See above. Why is the initial underbar used?

Reply:
See my response above - it's because I asked for them to be changed. I remember now that we stopped using this pattern so agree let's remove the leading underscore from all the method names in this example.